File: src/geek time/c#/week03/50. Pow(x, n).cs
using System;

namespace week03 {
    public partial class Solution {
        public double MyPow(double x, int n) {
        if (n < 0) {
            // 这一步也很好的解决了int.MinValue的问题， -2147483648 取反 得 0
            n = -n;
            x = 1 / x;
        }
        return Pow(x, n);
    }
        private double Pow(double x, int n) {
            if (n == 0) return 1d;
            var half = Pow(x, n /2);
            return n % 2 == 0 ? half * half : half * half * x;
        }
    }
}

File: src/geek time/python/week03/50. Pow(x, n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            n = -n
            x = 1 / x
        return self.recursivePow(x, n)
    def recursivePow(self, x: float, n: int):
        if n == 0:
            return 1
        half = self.recursivePow(x, n / 2)
        if n % 2 == 0:
            return half * half
        else:
            return half * half * x
        